src/goose/synopsis/system.py

```diff
@@ -0,0 +1,149 @@
+
+    def _cleanup_processes(self) -> None:
+        """Terminate all processes in the process queue."""
+        for process_id in list(self._processes.keys()):
+            self.cancel_process(process_id)
+
+    def add_process(self, process: subprocess.Popen) -> int:
+        """Add a new background process and return its assigned ID."""
+        process_id = process.pid
+        os.set_blocking(process.stdout.fileno(), False)
+        self._processes[process_id] = process
+        return process_id
+
+    def get_processes(self) -> Dict[int, str]:
+        """List all background processes with their IDs and commands."""
+        return {pid: str(proc.args) for pid, proc in self._processes.items()}
+
+    def view_process_output(self, process_id: int) -> str:
+        """View the output of a running background process."""
+        process = self._processes.get(process_id)
+        if not process:
+            raise ValueError(f"No process found with ID: {process_id}")
+
+        output = []
+        while line := process.stdout.readline():
+            output.append(line)
+
+        return "".join(output)
+
+    def cancel_process(self, process_id: int) -> bool:
+        """Cancel the background process with the specified ID."""
+        process = self._processes.pop(process_id, None)
+        if process:
+            process.terminate()
+            return True
+        return False
+
+    def to_relative(self, path: str) -> str:
+        """convert string path to be relative to cwd, assuming it starts absolute"""
+        return os.path.relpath(path, start=self.cwd)
+
+    def to_patho(self, path: str) -> Path:
+        """convert string to pathlib.Path object
+
+        This attempts to resolve the path against the cwd
+        """
+        patho = Path(path)
+        if patho.is_absolute():
+            return patho
+
+        return (self.cwd / patho).resolve()
+
+    def remember_file(self, path: str) -> None:
+        """Place a file into the active files"""
+        path = str(self.to_patho(path))
+
+        # Do a size check on the file to ensure we don't overload the LLM context
+        with open(path, "r") as f:
+            content = f.read()
+
+        max_output_chars = 2**20
+        max_output_tokens = 16000
+        encoder = get_encoding("cl100k_base")
+
+        if len(content) > max_output_chars or len(encoder.encode(content)) > max_output_tokens:
+            raise ValueError(f"The file at {path} is too large to read directly!")
+
+        self._active_files.add(path)
+
+    def forget_file(self, path: str) -> None:
+        """Forget an existing active file"""
+        self._active_files.discard(str(self.to_patho(path)))
+
+    def info(self) -> str:
+        """Summarize the current operating system"""
+        return json.dumps(
+            dict(
+                os=self.platform,
+                cwd=str(self.to_patho(self.cwd)),
+                shell=os.environ.get("SHELL", "unknown"),
```

Comment:
:nit: probably not the right place but this might return the incorrect shell. when i drop into a subshell i'll only get my login shell rather than the current shell execution environment. 

```
fish-3.7.1$ echo $SHELL
/opt/homebrew/bin/fish

fish-3.7.1$ bash
bash-5.2$ echo $SHELL
/opt/homebrew/bin/fish

bash-5.2$ python3 get-shell.py
bash
```

we should eventually try to consider this as a util

```python3
def get_current_shell():
    cmd = ["ps", "-p", str(subprocess.os.getppid()), "-o", "comm="]
    result = subprocess.run(cmd, capture_output=True, text=True)
    return result.stdout.lstrip('-') 
```

for example: this can sometimes cause quirks since in `fish` the `cd` is a function (as opposed to a `builtin <cmd>`) or because the BSD commands look the same but don't have all the same options/flags as with GNU.